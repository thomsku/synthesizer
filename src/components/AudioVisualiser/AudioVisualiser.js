import React, {Component} from "react";
import PropTypes from "prop-types";

class AudioVisualiser extends Component {
  constructor(props) {
    super(props);
    this.canvas = React.createRef();
  }

  componentDidUpdate() {
    this.draw();
  }

  draw() {
    const canvas = this.canvas.current;
    const width = canvas.width;
    const height = canvas.height;
    const context = canvas.getContext("2d");
    const scaling = height / 256;
    let risingEdge = 0;
    const edgeThreshold = 5;

    context.fillStyle = "rgba(40, 56, 68, 0.21)";
    context.clearRect(0, 0, width, height);

    context.lineWidth = 2;
    context.strokeStyle = "#fff";
    context.beginPath();

    // No buffer overrun protection
    while (this.props.audioData[risingEdge++] - 128 > 0 && risingEdge <= width);
    if (risingEdge >= width) risingEdge = 0;

    while (this.props.audioData[risingEdge++] - 128 < edgeThreshold && risingEdge <= width);
    if (risingEdge >= width) risingEdge = 0;

    for (let x = risingEdge; x < this.props.audioData.length && x - risingEdge < width; x++) {
      context.lineTo(x - risingEdge, height - this.props.audioData[x] * scaling);
    }

    context.stroke();
  }

  render() {
    return <canvas width="800" height="256" ref={this.canvas} />;
  }
}

AudioVisualiser.propTypes = {
  audioData: PropTypes.object.isRequired,
};

export default AudioVisualiser;
