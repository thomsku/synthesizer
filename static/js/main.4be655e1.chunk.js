(window.webpackJsonp=window.webpackJsonp||[]).push([[0],{135:function(e,a,t){e.exports=t(230)},140:function(e,a,t){},141:function(e,a,t){},155:function(e,a,t){},225:function(e,a,t){},227:function(e,a,t){},229:function(e,a,t){},230:function(e,a,t){"use strict";t.r(a);var n=t(0),i=t.n(n),l=t(23),s=t.n(l),o=(t(140),t(49)),r=t(88),c=t(28),u=t(17),d=t(31),h=t(29),m=t(9),v=t(32),p=function(e){function a(e){var t;return Object(c.a)(this,a),(t=Object(d.a)(this,Object(h.a)(a).call(this,e))).canvas=i.a.createRef(),t}return Object(v.a)(a,e),Object(u.a)(a,[{key:"componentDidUpdate",value:function(){this.draw()}},{key:"draw",value:function(){var e=this.canvas.current,a=e.width,t=e.height,n=e.getContext("2d"),i=t/256,l=0;for(n.fillStyle="rgba(40, 56, 68, 0.21)",n.clearRect(0,0,a,t),n.lineWidth=2,n.strokeStyle="#fff",n.beginPath();this.props.audioData[l++]-128>0&&l<=a;);for(l>=a&&(l=0);this.props.audioData[l++]-128<5&&l<=a;);l>=a&&(l=0);for(var s=l;s<this.props.audioData.length&&s-l<a;s++)n.lineTo(s-l,t-this.props.audioData[s]*i);n.stroke()}},{key:"render",value:function(){return i.a.createElement("canvas",{width:"800",height:"256",ref:this.canvas})}}]),a}(n.Component),f=function(e){function a(e){var t;return Object(c.a)(this,a),(t=Object(d.a)(this,Object(h.a)(a).call(this,e))).state={audioData:new Uint8Array(0)},t.tick=t.tick.bind(Object(m.a)(t)),t}return Object(v.a)(a,e),Object(u.a)(a,[{key:"componentDidMount",value:function(){this.analyser=this.props.audioContext.createAnalyser(),this.dataArray=new Uint8Array(this.analyser.frequencyBinCount),this.props.amp.connect(this.analyser),this.rafId=requestAnimationFrame(this.tick)}},{key:"tick",value:function(){this.analyser.getByteTimeDomainData(this.dataArray),this.setState({audioData:this.dataArray}),this.rafId=requestAnimationFrame(this.tick)}},{key:"componentWillUnmount",value:function(){cancelAnimationFrame(this.rafId),this.analyser.disconnect()}},{key:"render",value:function(){return i.a.createElement("div",null,i.a.createElement(p,{audioData:this.state.audioData}))}}]),a}(i.a.Component),b=t(252),y=t(254),g=t(250),E=t(253),k=(t(141),function(e){var a=e.setTuning,t=e.tuning;return i.a.createElement("div",{className:"radiobuttons"},i.a.createElement(E.a,{className:"parametertitle",component:"legend"},"Tuning"),i.a.createElement(y.a,{"aria-label":"Tuning",name:"tuning",className:"radiogroup",value:t,onChange:a},i.a.createElement(g.a,{className:"radiolabel",value:"just",control:i.a.createElement(b.a,{className:"radio"}),label:"just intonation"}),i.a.createElement(g.a,{className:"radiolabel",value:"equal",control:i.a.createElement(b.a,{className:"radio"}),label:"equal temperament"})))}),O=(t(155),function(e){var a=e.setWaveform,t=e.waveform;return i.a.createElement("div",{className:"radiobuttons"},i.a.createElement(E.a,{className:"parametertitle",component:"legend"},"Waveform"),i.a.createElement(y.a,{"aria-label":"Waveform",name:"waveform",className:"radiogroup",value:t,onChange:a},i.a.createElement(g.a,{className:"radiolabel",value:"sine",control:i.a.createElement(b.a,{className:"radio"}),label:"sine"}),i.a.createElement(g.a,{className:"radiolabel",value:"square",control:i.a.createElement(b.a,{className:"radio"}),label:"square"}),i.a.createElement(g.a,{className:"radiolabel",value:"sawtooth",control:i.a.createElement(b.a,{className:"radio"}),label:"sawtooth"}),i.a.createElement(g.a,{className:"radiolabel",value:"triangle",control:i.a.createElement(b.a,{className:"radio"}),label:"triangle"})))}),w=t(126),j=t.n(w),x=t(251),C=t(231),N=t(7),D=t(66),T=(t(225),Object(N.a)({thumb:{height:20,width:20,backgroundColor:"#fff","&$focused, &:hover":{boxShadow:"0px 0px 0px ".concat(8,"px ",Object(D.fade)("#de235b",.16))},"&$activated":{boxShadow:"0px 0px 0px ".concat(12,"px ").concat(Object(D.fade)("#de235b",.16))},"&$jumped":{boxShadow:"0px 0px 0px ".concat(12,"px ").concat(Object(D.fade)("#de235b",.16))}},track:{backgroundColor:"#fff",height:2},trackAfter:{backgroundColor:"#fff"},focused:{},activated:{},jumped:{}})(x.a)),A=function(e){function a(e){var t;return Object(c.a)(this,a),(t=Object(d.a)(this,Object(h.a)(a).call(this,e))).handleChange=t.handleChange.bind(Object(m.a)(t)),t}return Object(v.a)(a,e),Object(u.a)(a,[{key:"handleChange",value:function(e,a,t){this.props.setEnvelope(t,a)}},{key:"render",value:function(){var e=this;return i.a.createElement("div",{className:"envelope"},i.a.createElement(E.a,{className:"envelopeparametertitle",component:"legend"},"Envelope"),i.a.createElement("div",{className:"graph"},i.a.createElement(j.a,{style:{height:"100%",width:"100%"},lineStyle:{strokeWidth:"1.5px"},a:this.props.envelope.attack,d:this.props.envelope.decay,s:this.props.envelope.sustain,r:this.props.envelope.release})),i.a.createElement("div",{className:"sliders"},i.a.createElement(C.a,{id:"label"},"attack"),i.a.createElement(T,{id:"attack",className:"slider",value:this.props.envelope.attack,min:.01,max:2,"aria-labelledby":"label",onChange:function(a,t){e.handleChange(a,t,"attack")}}),i.a.createElement(C.a,{id:"label"},"decay"),i.a.createElement(T,{id:"attack",className:"slider",value:this.props.envelope.decay,min:.01,max:2,"aria-labelledby":"label",onChange:function(a,t){e.handleChange(a,t,"decay")}}),i.a.createElement(C.a,{id:"label"},"sustain"),i.a.createElement(T,{id:"attack",className:"slider",value:this.props.envelope.sustain,min:0,max:1,"aria-labelledby":"label",onChange:function(a,t){e.handleChange(a,t,"sustain")}}),i.a.createElement(C.a,{id:"label"},"release"),i.a.createElement(T,{id:"attack",className:"slider release",value:this.props.envelope.release,min:.01,max:2,"aria-labelledby":"label",onChange:function(a,t){e.handleChange(a,t,"release")}})))}}]),a}(i.a.Component),q={a:261.63,s:294.33,d:327.03,f:392.44,g:436.05,h:523.25,j:588.66,k:654.06,l:697.66},S={a:261.63,s:293.66,d:329.63,f:392,g:440,h:523.25,j:587.32,k:659.26,l:698.46},W=(t(227),t(228)),U=function(e){function a(e){var t;return Object(c.a)(this,a),(t=Object(d.a)(this,Object(h.a)(a).call(this,e))).state={activeOscillators:{},audioData:new Uint8Array(0),tuning:{type:"just",frequencies:q},waveform:"sine",envelope:{attack:.1,decay:.01,sustain:1,release:.1}},t.handleKeyDown=t.handleKeyDown.bind(Object(m.a)(t)),t.handleKeyUp=t.handleKeyUp.bind(Object(m.a)(t)),t.play=t.play.bind(Object(m.a)(t)),t.stop=t.stop.bind(Object(m.a)(t)),t.setTuning=t.setTuning.bind(Object(m.a)(t)),t.setWaveform=t.setWaveform.bind(Object(m.a)(t)),t.setEnvelope=t.setEnvelope.bind(Object(m.a)(t)),t.audioContext=new(window.AudioContext||window.webkitAudioContext),t.volume=t.audioContext.createGain(),t}return Object(v.a)(a,e),Object(u.a)(a,[{key:"componentDidMount",value:function(){this.volume.gain.value=.2,this.volume.connect(this.audioContext.destination),document.addEventListener("keydown",this.handleKeyDown),document.addEventListener("keyup",this.handleKeyUp)}},{key:"handleKeyDown",value:function(e){var a=e.key;!this.state.activeOscillators[a]&&this.state.tuning.frequencies[a]&&this.play(a)}},{key:"handleKeyUp",value:function(e){var a=e.key;this.state.activeOscillators[a]&&this.stop(a)}},{key:"play",value:function(e){var a=this.audioContext.createOscillator();a.frequency.value=this.state.tuning.frequencies[e],a.type=this.state.waveform;var t=this.audioContext.createGain();t.connect(this.volume);var n=new W(this.audioContext,t.gain);n.mode="ADSR",n.attackTime=this.state.envelope.attack,n.decayTime=this.state.envelope.decay,n.sustainLevel=this.state.envelope.sustain,n.releaseTime=this.state.envelope.release,this.setState(function(t){return{activeOscillators:Object(r.a)({},t.activeOscillators,Object(o.a)({},e,{oscillator:a,envelope:n}))}}),this.state.activeOscillators[e].oscillator.connect(t),this.state.activeOscillators[e].oscillator.start(),this.state.activeOscillators[e].envelope.gateOn(this.audioContext.currentTime)}},{key:"stop",value:function(e){this.state.activeOscillators[e].envelope.gateOff(this.audioContext.currentTime),this.state.activeOscillators[e].oscillator.stop(this.audioContext.currentTime+10),delete this.state.activeOscillators[e]}},{key:"setTuning",value:function(e){var a=q;"equal"===e.target.value&&(a=S),this.setState({tuning:{type:e.target.value,frequencies:a}})}},{key:"setWaveform",value:function(e){this.setState({waveform:e.target.value})}},{key:"setEnvelope",value:function(e,a){this.setState(function(t){return{envelope:Object(r.a)({},t.envelope,Object(o.a)({},e,a))}})}},{key:"render",value:function(){return i.a.createElement("div",{className:"synthDiv"},i.a.createElement(f,{audioContext:this.audioContext,activeOscillators:this.state.activeOscillators,amp:this.volume}),i.a.createElement("h2",null,"Synth Parameters"),i.a.createElement("div",{className:"parameters"},i.a.createElement("div",{className:"radioparameters"},i.a.createElement(k,{setTuning:this.setTuning,tuning:this.state.tuning.type}),i.a.createElement(O,{setWaveform:this.setWaveform,waveform:this.state.waveform})),i.a.createElement(A,{setEnvelope:this.setEnvelope,envelope:this.state.envelope})))}}]),a}(i.a.Component);t(229);var K=function(){return i.a.createElement("div",{className:"App"},i.a.createElement(U,null))};Boolean("localhost"===window.location.hostname||"[::1]"===window.location.hostname||window.location.hostname.match(/^127(?:\.(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}$/));s.a.render(i.a.createElement(K,null),document.getElementById("root")),"serviceWorker"in navigator&&navigator.serviceWorker.ready.then(function(e){e.unregister()})}},[[135,1,2]]]);
//# sourceMappingURL=main.4be655e1.chunk.js.map