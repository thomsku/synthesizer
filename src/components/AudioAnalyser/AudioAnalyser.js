import React from "react";
import AudioVisualiser from "../AudioVisualiser/AudioVisualiser";

class AudioAnalyser extends React.Component {
  constructor(props) {
    super(props);

    this.state = {audioData: new Uint8Array(0)};
  }
  render() {
    return (
      <div>
        <AudioVisualiser audioData={this.state.audioData}/>
      </div>
    );
  }
}

export default AudioAnalyser;
