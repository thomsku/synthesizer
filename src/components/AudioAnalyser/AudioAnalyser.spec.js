import React from "react";
import {shallow, mount} from "enzyme";
import AudioAnalyser from "./AudioAnalyser";
import AudioVisualiser from "../AudioVisualiser/AudioVisualiser";
require("../Synth/AudioContext.mock.js");

describe("AudioAnalyser", () => {
  let wrapper;
  const audioContext = new (window.AudioContext || window.webkitAudioContext)();
  const amp = audioContext.createGain();
  beforeEach(() => wrapper = shallow(<AudioAnalyser audioContext={audioContext} activeOscillators={{}} amp={amp}/>));

  it("should render correctly", () => expect(wrapper).toMatchSnapshot());

  it("should render a <div />", () => {
    expect(wrapper.find("div")).toHaveLength(1);
  });

  it("should render the AudioVisualiser Component", () => {
    expect(wrapper.containsMatchingElement(<AudioVisualiser audioData={wrapper.instance().state.audioData}/>)).toEqual(true);
  });
});

describe("tick", () => {
  let wrapper;
  window.HTMLCanvasElement.prototype.getContext = () => {
    return {
      width: 0,
      beginPath: jest.fn(),
      clearRect: jest.fn(),
      lineTo: jest.fn(),
      moveTo: jest.fn(),
      stroke: jest.fn(),
    };
  };
  const audioContext = new (window.AudioContext || window.webkitAudioContext)();
  const amp = audioContext.createGain();
  beforeEach(() => {
    wrapper = mount(<AudioAnalyser audioContext={audioContext} activeOscillators={{}} amp={amp}/>);
    jest.spyOn(window, "requestAnimationFrame").mockImplementationOnce((cb) => cb());
  });

  it("calls tick() when the component is mounted", () => {
    const spy = jest.spyOn(wrapper.instance(), "tick");
    wrapper.instance().componentDidMount();
    expect(spy).toHaveBeenCalledTimes(1);
  });

  afterEach(() => {
    window.requestAnimationFrame.mockRestore();
  });
});
