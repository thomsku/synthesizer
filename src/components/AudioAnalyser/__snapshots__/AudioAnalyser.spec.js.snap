// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioAnalyser should render correctly 1`] = `
<div>
  <AudioVisualiser
    audioData={Uint8Array []}
  />
</div>
`;
